[CSS] get(Bounding)ClientRect API returns rects scaled lengths

mdubet opened the pull request:
#### 4e36d89dbce0a0b20d598724f9fcdee5350fec32
<pre>
[CSS] get(Bounding)ClientRect API returns rects scaled lengths
<a href="https://bugs.webkit.org/show_bug.cgi?id=300474">https://bugs.webkit.org/show_bug.cgi?id=300474</a>
<a href="https://rdar.apple.com/162325730">rdar://162325730</a>

Reviewed by Sammy Gill.

The behavior of this API has changed, it now returns scaled length instead of unscaled ones.
Most of this patch is updating tests to account for this new behavior (by dividing by the CSS zoom).

<a href="https://drafts.csswg.org/css-viewport/#scaled">https://drafts.csswg.org/css-viewport/#scaled</a>
<a href="https://drafts.csswg.org/css-viewport/#zoom-om">https://drafts.csswg.org/css-viewport/#zoom-om</a>

* LayoutTests/TestExpectations:
* LayoutTests/css3/zoom-coords.xhtml:
* LayoutTests/fast/block/float/floats-with-margin-should-not-wrap.html:
* LayoutTests/fast/css/resize-below-min-intrinsic-size.html:
* LayoutTests/fast/css/resize-below-min-size-zoomed.html:
* LayoutTests/fast/dom/Document/CaretRangeFromPoint/caretRangeFromPoint-in-zoom-and-scroll.html:
* LayoutTests/fast/forms/search-zoomed.html:
* LayoutTests/fast/forms/switch/zoom-computed-style.html:
* LayoutTests/fast/sub-pixel/layout-boxes-with-zoom-expected.html:
* LayoutTests/fast/sub-pixel/layout-boxes-with-zoom.html:
* LayoutTests/fast/sub-pixel/table-rows-have-stable-height.html:
* LayoutTests/fast/transforms/bounding-rect-zoom-expected.txt: Added.
* LayoutTests/fast/transforms/bounding-rect-zoom.html:
* LayoutTests/fast/zooming/client-rects-with-css-and-page-zoom-expected.txt:
* LayoutTests/fast/zooming/client-rects-with-css-and-page-zoom.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-viewport/zoom/relative-units-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-viewport/zoom/widget-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/cssom-view/getBoundingClientRect-zoom-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/cssom-view/getClientRects-zoom-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/cssom-view/image-x-y-zoom-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/cssom-view/scrollIntoView-inline-image-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/cssom-view/scrollIntoView-sideways-lr-writing-mode-and-rtl-direction-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/cssom-view/scrollIntoView-sideways-lr-writing-mode-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/intersection-observer/bounding-box-expected.txt:
* LayoutTests/platform/glib/fast/transforms/bounding-rect-zoom-expected.txt: Removed.
* LayoutTests/platform/gtk/fast/transforms/bounding-rect-zoom-expected.png: Removed.
* LayoutTests/platform/ios/fast/transforms/bounding-rect-zoom-expected.png: Removed.
* LayoutTests/platform/ios/fast/transforms/bounding-rect-zoom-expected.txt: Removed.
* LayoutTests/platform/mac/css3/zoom-coords-expected.txt:
* LayoutTests/platform/mac/fast/transforms/bounding-rect-zoom-expected.png: Removed.
* LayoutTests/platform/mac/fast/transforms/bounding-rect-zoom-expected.txt: Removed.
* LayoutTests/platform/win/fast/transforms/bounding-rect-zoom-expected.txt: Removed.
* LayoutTests/svg/zoom/page/zoom-zoom-coords-expected.txt:
* LayoutTests/svg/zoom/page/zoom-zoom-coords.xhtml:
* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::zoomForClient const):
(WebCore::Document::convertAbsoluteToClientQuads):
(WebCore::Document::convertAbsoluteToClientRects):
(WebCore::Document::convertAbsoluteToClientRect):
* Source/WebCore/dom/Document.h:

Canonical link: <a href="https://commits.webkit.org/301806@main">https://commits.webkit.org/301806@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/764029f0ac88ed41c49c7ecede0c5cd309720b1f

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/126999 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/46635 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/37639 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/134003 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/78564 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/ce98c1fb-469a-41d7-b8d6-651207474c1f) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/128870 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/47252 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/55162 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/96638 "Passed tests") | [❌ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/64643 "") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/8ec2354b-6d8e-44a7-80da-11e5cc4ef811) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/129947 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/37814 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/113612 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/77149 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/74d2471c-574f-44e2-8383-2e246e930972) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/36677 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/31766 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/77395 "Built successfully") | | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/119036 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/107658 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/32096 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/136529 "Built successfully") | | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/125462 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/53654 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/41314 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/105155 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/54158 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/109972 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/104846 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/26759 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/50366 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/28722 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/51160 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/53586 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/59459 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/158500 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/52825 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/39647 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/56159 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/54584 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

sammygill commented on Source/WebCore/dom/Document.cpp:
```diff
@@ -9593,13 +9579,23 @@ Element* eventTargetElementForDocument(Document* document)
     return element;
 }
 
+// get(Bounding)ClientRect APIs now returns scaled (=zoomed) rect.
+// The zoom argument will be used to divide the rect, returning an unzoomed rect if passed.
+// https://drafts.csswg.org/css-viewport/#zoom-om
```
It might be nice to elaborate on this comment a bit more to explain *why* we need to do something like this because I don't think it's obvious from just looking at it.

I think that will also make the implications of the second part of this change more explicit - that we choose want to return `std::nullopt` so that the the code for determining the rect doesn't remove the scales.

mdubet committed:
[CSS] get(Bounding)ClientRect API returns rects scaled lengths
https://bugs.webkit.org/show_bug.cgi?id=300474
rdar://162325730

Reviewed by Sammy Gill.

The behavior of this API has changed, it now returns scaled length instead of unscaled ones.
Most of this patch is updating tests to account for this new behavior (by dividing by the CSS zoom).

https://drafts.csswg.org/css-viewport/#scaled
https://drafts.csswg.org/css-viewport/#zoom-om

* LayoutTests/TestExpectations:
* LayoutTests/css3/zoom-coords.xhtml:
* LayoutTests/fast/block/float/floats-with-margin-should-not-wrap.html:
* LayoutTests/fast/css/resize-below-min-intrinsic-size.html:
* LayoutTests/fast/css/resize-below-min-size-zoomed.html:
* LayoutTests/fast/dom/Document/CaretRangeFromPoint/caretRangeFromPoint-in-zoom-and-scroll.html:
* LayoutTests/fast/forms/search-zoomed.html:
* LayoutTests/fast/forms/switch/zoom-computed-style.html:
* LayoutTests/fast/sub-pixel/layout-boxes-with-zoom-expected.html:
* LayoutTests/fast/sub-pixel/layout-boxes-with-zoom.html:
* LayoutTests/fast/sub-pixel/table-rows-have-stable-height.html:
* LayoutTests/fast/transforms/bounding-rect-zoom-expected.txt: Added.
* LayoutTests/fast/transforms/bounding-rect-zoom.html:
* LayoutTests/fast/zooming/client-rects-with-css-and-page-zoom-expected.txt:
* LayoutTests/fast/zooming/client-rects-with-css-and-page-zoom.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-viewport/zoom/relative-units-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-viewport/zoom/widget-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/cssom-view/getBoundingClientRect-zoom-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/cssom-view/getClientRects-zoom-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/cssom-view/image-x-y-zoom-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/cssom-view/scrollIntoView-inline-image-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/cssom-view/scrollIntoView-sideways-lr-writing-mode-and-rtl-direction-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/cssom-view/scrollIntoView-sideways-lr-writing-mode-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/intersection-observer/bounding-box-expected.txt:
* LayoutTests/platform/glib/fast/transforms/bounding-rect-zoom-expected.txt: Removed.
* LayoutTests/platform/gtk/fast/transforms/bounding-rect-zoom-expected.png: Removed.
* LayoutTests/platform/ios/fast/transforms/bounding-rect-zoom-expected.png: Removed.
* LayoutTests/platform/ios/fast/transforms/bounding-rect-zoom-expected.txt: Removed.
* LayoutTests/platform/mac/css3/zoom-coords-expected.txt:
* LayoutTests/platform/mac/fast/transforms/bounding-rect-zoom-expected.png: Removed.
* LayoutTests/platform/mac/fast/transforms/bounding-rect-zoom-expected.txt: Removed.
* LayoutTests/platform/win/fast/transforms/bounding-rect-zoom-expected.txt: Removed.
* LayoutTests/svg/zoom/page/zoom-zoom-coords-expected.txt:
* LayoutTests/svg/zoom/page/zoom-zoom-coords.xhtml:
* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::zoomForClient const):
(WebCore::Document::convertAbsoluteToClientQuads):
(WebCore::Document::convertAbsoluteToClientRects):
(WebCore::Document::convertAbsoluteToClientRect):
* Source/WebCore/dom/Document.h:

Canonical link: https://commits.webkit.org/301806@main
